packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/InAppPurchasePlugin.swift

```diff
@@ -0,0 +1,425 @@
+        details: paymentMap)
+      return
+    }
+
+    guard let product = self.getProduct(productID: productID) else {
+      error.pointee = FlutterError(
+        code: "storekit_invalid_payment_object",
+        message:
+          "You have requested a payment for an invalid product. Either the `productIdentifier` of the payment is not valid or the product has not been fetched before adding the payment to the payment queue.",
+        details: paymentMap)
+      return
+    }
+
+    let payment = SKMutablePayment(product: product)
+    payment.applicationUsername = paymentMap["applicationUsername"] as? String
+    payment.quantity = paymentMap["quantity"] as? Int ?? 1
+    payment.simulatesAskToBuyInSandbox = paymentMap["simulatesAskToBuyInSandbox"] as? Bool ?? false
+
+    if #available(iOS 12.2, *) {
+      if let paymentDiscountMap = paymentMap["paymentDiscount"] as? [String: Any],
+        !paymentDiscountMap.isEmpty
+      {
+        var invalidError: NSString?
+        if let paymentDiscount = FIAObjectTranslator.getSKPaymentDiscount(
+          fromMap: paymentDiscountMap, withError: &invalidError)
+        {
+          payment.paymentDiscount = paymentDiscount
+        } else if let invalidError = invalidError {
+          error.pointee = FlutterError(
+            code: "storekit_invalid_payment_discount_object",
+            message:
+              "You have requested a payment and specified a payment discount with invalid properties. \(invalidError)",
+            details: paymentMap)
+          return
+        }
+      }
+    }
+
+    if !self.paymentQueueHandler!.add(payment) {
+      error.pointee = FlutterError(
+        code: "storekit_duplicate_product_object",
+        message:
+          "There is a pending transaction for the same product identifier. Please either wait for it to be finished or finish it manually using `completePurchase` to avoid edge cases.",
+        details: paymentMap)
+      return
+    }
+  }
+
+  public func finishTransactionFinishMap(
+    _ finishMap: [String: Any], error: AutoreleasingUnsafeMutablePointer<FlutterError?>
+  ) {
+    let transactionIdentifier = (finishMap["transactionIdentifier"] ?? "") as? String
+    guard let productIdentifier = finishMap["productIdentifier"] as? String
+    else {
+      return
+    }
+
+    let pendingTransactions = paymentQueueHandler!.getUnfinishedTransactions()
+
+    for transaction in pendingTransactions {
+      // If the user cancels the purchase dialog we won't have a transactionIdentifier.
+      // So if it is null AND a transaction in the pendingTransactions list has
+      // also a null transactionIdentifier we check for equal product identifiers.
+      if transaction.transactionIdentifier == transactionIdentifier
+        || (transactionIdentifier == ""
+          && transaction.transactionIdentifier == nil
+          && transaction.payment.productIdentifier == productIdentifier)
+      {
+        paymentQueueHandler!.finish(transaction)
```

Comment:
Writing `!` all over the code makes it pretty hard to reason about, and can easily be overlooked, defeating the purpose of null safety. If this is guaranteed to be non nil, it's more readable to move the cast to the front and explain why:  

```
guard let paymentQueueHandler else { 
  fatalError("paymentQueueHandler cant be nil because reason 1 2 3") 
}
```